### ISSUE TITLE
A9: Model Training & Hyperparameter Tuning

### ISSUE THREAD

**Issue by amolk99** [othmanreem/Data-intensive-systems#37]
## Objective
Train Deep Learning models with various hyperparameters and implement cross-validation and early stopping strategies.

### Model Compilation
- [ ] Test optimizer variants:
  - SGD (Stochastic Gradient Descent)
  - RMSprop
  - Adam
  - Document learning rate settings for each
- [ ] Configure loss functions for regression:
  - MSE (Mean Squared Error)
  - MAE (Mean Absolute Error)
  - Test and compare both
- [ ] Add evaluation metrics:
  - Include MAE for model quality assessment
  - Track metrics in training history

### Training Configuration
- [ ] Implement varying number of epochs:
  - Test range: 10, 50, 100, 200, 500
  - Document convergence patterns
- [ ] Implement batch size variations:
  - Test batch sizes: 16, 32, 64, 128
- [ ] Track and log training history (loss, metrics per epoch)

### Early Stopping
- [ ] Split validation set from training data
- [ ] Implement early stopping callback:
  ```python
  tf.keras.callbacks.EarlyStopping(
      monitor='val_loss',
      patience=10,
      restore_best_weights=True
  )
  ```
- [ ] Document early stopping effectiveness
- [ ] Compare models with/without early stopping

### Cross-Validation
- [ ] Implement K-fold cross-validation strategy
- [ ] Use ~10 random sequences per fold for testing (per instructions)
- [ ] Calculate mean and std of validation metrics across folds
- [ ] Document cross-validation procedure

### Hyperparameter Grid Search
- [ ] Create systematic hyperparameter combinations:
  ```python
  params = {
      'optimizer': ['sgd', 'adam', 'rmsprop'],
      'layers': [1, 2, 3],
      'nodes': [32, 64, 128],
      'epochs': [50, 100, 200],
      'batch_size': [32, 64]
  }
  ```
- [ ] Log all experiment results
- [ ] Identify top-performing configurations
- [ ] Document combinations that failed (dead ends)

### Training Logs
- [ ] Save training history for each configuration
- [ ] Create visualization of loss curves
- [ ] Export results for reporting

## Acceptance Criteria
- Multiple optimizer configurations tested and compared
- Early stopping implemented and validated
- Cross-validation with 10 test sequences per fold
- Hyperparameter search completed with results logged
- Training history saved for all experiments
- Dead ends documented (configurations that didn't work)

## Dependencies
- Requires: Issue #36 (Data Loading & Neural Network Architecture)

## Related Files
- Input: Models from Issue #36
- Output: `A9/train.py`
- Output: `A9/training_logs/`
